### PR TITLE
Script for integrating PRs

### PR DESCRIPTION
**Pull request by slaskawi**
This scripts helps integrating pull requests in "cherry-pick" fashion.

At first, run this command in your Infinispan project directory: `./integrate_pr.sh --setup -r upstream` (assuming that Infinispan's remote is called `upstream`)
Then integrate PR using: `./integrate_pr.sh -db master -r upstream -pr 332` (assuming that Infinispan's remote is called upstream, destination branch is called `master` and we'd like to integrate Pull Request number `332`).

Any comments highly appreciated.

### CODE REVIEW

**Comment by danberindei** [comment]
I don't think this should be in the infinispan repository, as the exact script you'll run will depend on the branch you're on. 

I also don't like the idea of integrating a PR without having at least a cursory look at it in the IDE and running the core tests on my machine. My workflow is to fetch a PR with [gh-fetch](https://github.com/danberindei/scripts/blob/master/gh-fetch), run the tests locally while I have a look at the changes in IntelliJ, and then integrating with `git push origin pr_branch:master`.

File: bin/integrate_pr.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+shift
+done
+
+if [ "x$SETUP" != "x" ] && [ "x$REMOTE" != "x" ]; then
+
+	echo "Setting up remote for Pull Request integration"
+	git config --replace-all remote.$REMOTE.fetch "+refs/heads/*:refs/remotes/$REMOTE/*"
+	git config --add remote.$REMOTE.fetch "+refs/pull/*:refs/remotes/$REMOTE/pr/*"
+	echo "Fetching new refspecs"
+	git fetch $REMOTE --prune
```
Comment (danberindei) [inline bin/integrate_pr.sh]: Why change the config and not fetch a single ref like this: `git fetch upstream refs/pull/1/head:bla`?

File: bin/integrate_pr.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+set -e
+
+echo "Rebasing $REMOTE/pr/$PULL_REQUEST/head onto $REMOTE/$DESTINATION_BRANCH"
+git rebase --onto $REMOTE/$DESTINATION_BRANCH $REMOTE/$DESTINATION_BRANCH $REMOTE/pr/$PULL_REQUEST/head
+git checkout -B $DESTINATION_BRANCH 
+echo "Pushing to REMOTE"
+git push $DRY_RUN $REMOTE
+echo "Removing merge branch: $REMOTE/pr/$PULL_REQUEST/merge"
+# This may fail if pushing above commits resulted in fast forward
+git push $DRY_RUN $REMOTE :$REMOTE/pr/$PULL_REQUEST/merge || true
```
Comment (danberindei) [inline bin/integrate_pr.sh]: Would this delete the `refs/pull/1/merge` ref in the upstream repository? Why?

**Comment by tristantarrant** [comment]
I'd like to see a cross between the two approaches: use the GitHub API to retrieve the open PRs, allow the user to choose the one to try out, checkout out an "integration" branch based on the PR's base (should also allow choosing the branch for backports), rebase the PR on top of it and allow the user to inspect it with the IDE / tests. Then automation of push, closing the PR and the Jira after confirmation

**Comment by slaskawi** [comment]
Needs different approach - closing

**Pull request closed without merging**